src/server/socket.js

```diff
@@ -1,103 +1,204 @@
+    });
 
-            // Set up live reload if necessary.
-            if (config.liveReload) {
-                log.log('Starting live reload.');
-                livereload.init(socket);
-            }
+    socket.on('telemetry', function (data) {
+        telemetry.handleClientTelemetry(data);
+    });
 
-            // Set up telemetry if necessary.
-            if (config.telemetry) {
-                socket.emit('init-telemetry');
-            }
-            
-            // Set up xhr proxy
-            if (config.xhrProxy) {
-                socket.emit('init-xhr-proxy');
-            }
+    // Set up live reload if necessary.
+    if (config.liveReload) {
+        log.log('Starting live reload.');
+        livereload.init(socket);
+    }
 
-            handlePendingEmits(appHost);
-        });
+    // Set up telemetry if necessary.
+    if (config.telemetry) {
+        socket.emit('init-telemetry');
+    }
+    
+    // Set up xhr proxy
+    if (config.xhrProxy) {
+        socket.emit('init-xhr-proxy');
+    }
 
-        socket.on('register-simulation-host', function () {
-            log.log('Simulation host registered with server.');
+    handlePendingEmits(APP_HOST);
+}
 
-            // It only makes sense to have one simulation host per server. If more than one tries to connect, always
-            // take the most recent.
-            hostSockets[simHost] = socket;
+function handleSimHostRegistration(socket) {
+    socket.once('ready', handleSimHostReady);
+    emitToHost(SIM_HOST, 'init');
+}
 
-            socket.on('exec-success', function (data) {
-                emitToHost(appHost, 'exec-success', data);
-            });
-            socket.on('exec-failure', function (data) {
-                emitToHost(appHost, 'exec-failure', data);
-            });
+function onAppHostConnected() {
+    var socket = hostSockets.APP_HOST;
+    socket.once('app-plugin-list', handleAppPluginList);
 
-            socket.on('plugin-message', function (data) {
-                emitToHost(appHost, 'plugin-message', data);
-            });
+    emitToHost(APP_HOST, 'init');
+}
 
-            socket.on('plugin-method', function (data, callback) {
-                emitToHost(appHost, 'plugin-method', data, callback);
-            });
+function onSimHostReady() {
+    setupAppHostHandlers();
+}
 
-            socket.on('telemetry', function (data) {
-                telemetry.handleClientTelemetry(data);
-            });
+function handleSimHostReady() {
+    // resolving this promise will result in app-host handlers setupp
+    whenSimHostReady.resolve();
+
+    setupSimHostHandlers();
+
+    whenAppHostConnected.promise
+        .then(onAppHostConnected);
+}
+
+function handleAppPluginList(data) {
+    var socket = hostSockets.SIM_HOST;
 
-            // Set up telemetry if necessary.
-            if (config.telemetry) {
-                socket.emit('init-telemetry');
+    socket.once('start', handleStart);
```

Comment:
This function is called even when there's no sim-host instance connected. Maybe this scenario is not taken into account yet. To reproduce it, launch the simulation, you will get sim-host and app-host connected, then close sim-host so it gets disconnected, and close app-host. Load app-host one more time, and you will si an exception in the server log: `TypeError: Cannot read property 'once' of undefined`.